service/matching/version_rule_helpers.go

```diff
@@ -0,0 +1,209 @@
+// - No more rules than dynamicconfig.VersionAssignmentRuleLimitPerQueue
+// - No assignment rule's TargetBuildId can be a member of an existing version set
+// - If there existed an "unfiltered" assigment rule (which can accept any build id), at least one must still exist
+// - To override the unfiltered assignment rule requirement, the user can specify force = true
+func checkAssignmentConditions(g *persistencepb.VersioningData, maxARs int, force, hadUnfiltered bool) error {
+	rules := slices.Clone(g.GetAssignmentRules())
+	if maxARs > 0 && countActiveAR(rules) > maxARs {
+		return serviceerror.NewFailedPrecondition(fmt.Sprintf("update exceeds number of assignment rules permitted in namespace dynamic config (%v/%v)", len(rules), maxARs))
+	}
+	if tbid, ok := isInVersionSet(rules, g.GetVersionSets()); ok {
+		return serviceerror.NewFailedPrecondition(fmt.Sprintf("update breaks requirement, target build id %s is already a member of version set", tbid))
+	}
+	if force == true {
+		return nil
+	}
+	if hadUnfiltered && !hasUnfiltered(rules) {
+		return serviceerror.NewFailedPrecondition("update breaks requirement that at least one assignment rule must have no ramp or hint")
+	}
+	return nil
+}
+
+func countActiveAR(rules []*persistencepb.AssignmentRule) int {
+	cnt := 0
+	for _, rule := range rules {
+		if rule.DeleteTimestamp == nil {
+			cnt++
+		}
+	}
+	return cnt
+}
+
+// hasUnfiltered returns true if there exists an assignment rule with no hint, no worker ratio ramp, and no ramp percentage, or a ramp percentage of 100
+func hasUnfiltered(rules []*persistencepb.AssignmentRule) bool {
+	isUnfiltered := func(ar *taskqueue.BuildIdAssignmentRule) bool {
+		percentageRamp := ar.GetPercentageRamp()
+		return ar.GetHintFilter() == "" &&
+			ar.GetWorkerRatioRamp() == nil &&
+			(percentageRamp == nil || (percentageRamp != nil && percentageRamp.RampPercentage == 100))
+	}
+	found := false
+	for _, rule := range rules {
+		ar := rule.GetRule()
+		if rule.DeleteTimestamp == nil && isUnfiltered(ar) {
+			found = true
+		}
+	}
+	return found
+}
+
+// isInVersionSet returns true if the target build id of any assignment rule is in any of the listed version sets
+func isInVersionSet(rules []*persistencepb.AssignmentRule, sets []*persistencepb.CompatibleVersionSet) (string, bool) {
+	for _, rule := range rules {
+		ar := rule.GetRule()
+		tbid := ar.GetTargetBuildId()
+		if rule.DeleteTimestamp == nil {
+			for _, set := range sets {
+				for _, bid := range set.BuildIds {
+					if bid.GetId() == tbid {
+						return tbid, true
+					}
+				}
+			}
+		}
+	}
+	return "", false
+}
+
+// given2ActualIdx takes in the user-given index, which only counts active assignment rules, and converts it to the
+// actual index of that rule in the assignment rule list, which includes deleted rules.
+// A negative return value means index out of bounds.
+func given2ActualIdx(idx int32, rules []*persistencepb.AssignmentRule) int {
```

Comment:
may be convertGivenToActual, or something with convert?

Reply:
i think the "to" implies conversion without needing to make a long function name

Reply:
especially since there's a function comment above that callers should read anyways